src/client/dfs/dfs_sys.c

```diff
@@ -0,0 +1,1438 @@
+/**
+ * (C) Copyright 2018-2021 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+#include <libgen.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/xattr.h>
+#include <linux/xattr.h>
+#include <daos/common.h>
+#include <daos/event.h>
+
+#include <gurt/atomic.h>
+
+#include "daos.h"
+#include "daos_fs.h"
+
+#include "daos_fs_sys.h"
+
+/** Number of entries for readdir */
+#define DFS_SYS_NUM_DIRENTS 24
+
+/** Size of the hash table */
+#define DFS_SYS_HASH_SIZE 16
```

Comment:
I've changed it to 12, which is 2^12=4096, since we do have test cases with 50k directories.
In the future, we could consider making this configurable, or at least get a better grasp on the performance.